<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Reference · ABC-MK</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">ABC-MK</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Package Overview</span><ul><li><a class="tocitem" href="../analytical/">Analytical estimations</a></li><li><a class="tocitem" href="../abc/">ABC inference</a></li></ul></li><li class="is-active"><a class="tocitem" href>Reference</a><ul class="internal"><li><a class="tocitem" href="#Model-parameters-1"><span>Model parameters</span></a></li><li class="toplevel"><a class="tocitem" href="#Analytical-estimation-1"><span>Analytical estimation</span></a></li><li><a class="tocitem" href="#Fixations-1"><span>Fixations</span></a></li><li><a class="tocitem" href="#Polymorphism-1"><span>Polymorphism</span></a></li><li><a class="tocitem" href="#Summary-statistics-1"><span>Summary statistics</span></a></li><li><a class="tocitem" href="#Inference-tools-1"><span>Inference tools</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Reference</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/jmurga/Analytical.jl/blob/master/docs/src/reference.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h2 id="Model-parameters-1"><a class="docs-heading-anchor" href="#Model-parameters-1">Model parameters</a><a class="docs-heading-anchor-permalink" href="#Model-parameters-1" title="Permalink"></a></h2><p><em>adap</em> is the only variable exported from <em>Analytical</em> module. It is a Mutable structure contaning the variables required to solve the analytical approach. Any value can be easly changed. Remember <em>adap</em> should be change before the execution, in other case, <span>$\alpha_{(x)}$</span> will be solve with the default values. To change all the values at once, you can use <a href="#Analytical.parameters"><code>Analytical.parameters</code></a> in order to set specific models in a new variable.</p><article class="docstring"><header><a class="docstring-binding" id="Analytical.parameters" href="#Analytical.parameters"><code>Analytical.parameters</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">parameters(
	gam_neg::Int64,
	gL::Int64,
	gH::Int64,
	alLow::Float64,
	alTot::Float64,
	theta_f::Float64,
	theta_mid_neutral::Float64,
	al::Float64,
	be::Float64,
	B::Float64,
	bRange::Array{Float64,1}
	pposL::Float64,
	pposH::Float64,
	N::Int64,
	n::Int64,
	Lf::Int64,
	rho::Float64,
	TE::Float64,
)</code></pre><p>Mutable structure containing the variables required to solve the analytical approach. All the functions are solve using the internal values of the structure. For this reason, <em>adap</em> is the only exported variable. Adap should be change before the perform the analytical approach, in other case, <code>$\alpha_{(x)}$</code> will be solve with the default values.</p><p><strong>Parameters</strong></p><ul><li><code>gam_neg::Int64</code>:</li><li><code>gL::Int64</code>:</li><li><code>gH::Int64</code>:</li><li><code>alLow::Float64</code>:</li><li><code>alTot::Float64</code>:</li><li><code>theta_f::Float64</code>:</li><li><code>theta_mid_neutral::Float64</code>:</li><li><code>al::Float64</code>:</li><li><code>be::Float64</code>:</li><li><code>B::Float64</code>:</li><li><code>bRange::Array{Float64,1}</code>:</li><li><code>pposL::Float64</code>:</li><li><code>pposH::Float64</code>:</li><li><code>N::Int64</code>:</li><li><code>n::Int64</code>:</li><li><code>Lf::Int64</code>:</li><li><code>rho::Float64</code>:</li><li><code>TE::Float64</code>:</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/JuliaLang/julia/blob/697e782ab86bfcdd7fd15550241fe162c51d9f98/base/#L0-L44">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.Br" href="#Analytical.Br"><code>Analytical.Br</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">Br(Lmax,theta)</code></pre><p>Expected reduction in nucleotide diversity. Explored at <a href="https://doi.org/10.1017/S0016672300032365">Charlesworth B., 1994</a>:</p><div>\[\frac{\pi}{\pi_{0}} = e^{\frac{4\muL}{2rL+t}}\]</div><p><strong>Arguments</strong></p><ul><li><code>Lmax::Int64</code>: non-coding flaking length</li><li><code>theta::Float64</code></li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected reduction in diversity given a non-coding length, mutation rate and defined recombination.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/parameters.jl#L82-L96">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.set_theta_f!" href="#Analytical.set_theta_f!"><code>Analytical.set_theta_f!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">set_theta_f!(param)</code></pre><p>Find the optimum mutation given the expected reduction in nucleotide diversity (B value) in a locus.</p><p><strong>Returns</strong></p><ul><li><code>adap.theta_f::Float64</code>: changes adap.theta_f value.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/parameters.jl#L109-L116">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>Analytical.setPpos!</code>. Check Documenter&#39;s build log for details.</p></div></div><article class="docstring"><header><a class="docstring-binding" id="Analytical.binomOp!" href="#Analytical.binomOp!"><code>Analytical.binomOp!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">binomOp(param)</code></pre><p>Site Frequency Spectrum convolution depeding on background selection values. Pass the SFS to a binomial distribution to sample the allele frequencies probabilites.</p><p><strong>Returns</strong></p><ul><li><code>Array{Float64,2}</code>: convoluted SFS given a B value. It will be saved at <em>adap.bn</em>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/parameters.jl#L167-L174">source</a></section></article><h1 id="Analytical-estimation-1"><a class="docs-heading-anchor" href="#Analytical-estimation-1">Analytical estimation</a><a class="docs-heading-anchor-permalink" href="#Analytical-estimation-1" title="Permalink"></a></h1><h2 id="Fixations-1"><a class="docs-heading-anchor" href="#Fixations-1">Fixations</a><a class="docs-heading-anchor-permalink" href="#Fixations-1" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Analytical.fixNeut" href="#Analytical.fixNeut"><code>Analytical.fixNeut</code></a> — <span class="docstring-category">Function</span></header><section><div><ul><li>fixNeut()</li></ul><p>Expected neutral fixations rate reduce by a background selection value.</p><div>\[\mathbb{E}[D_{s}] = (1 - p_{-} - p_{+}) B \frac{1}{2N}\]</div><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of neutral fixations.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/fixations.jl#L9-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.fixNegB" href="#Analytical.fixNegB"><code>Analytical.fixNegB</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">fixNegB(ppos)</code></pre><p>Expected fixation rate from negative DFE.</p><div>\[\mathbb{E}[D_{n-}] =  p_{-}\left(2^-\alpha\beta^\alpha\left(-\zeta[\alpha,\frac{2+\beta}{2}] + \zeta[\alpha,1/2(2-\frac{1}{N+\beta})]\right)\right)\]</div><p><strong>Arguments</strong></p><ul><li><code>ppos::Float64</code>: positive selected alleles probabilty.</li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of fixations from negative DFE.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/fixations.jl#L30-L46">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.pFix" href="#Analytical.pFix"><code>Analytical.pFix</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">pFix()</code></pre><p>Expected positive fixation rate.</p><div>\[\mathbb{E}[D_{n+}] =  p_{+} \cdot B \cdot (1 - e^{(-2s)})\]</div><p><strong>Arguments</strong></p><ul><li><code>ppos::Float64</code>: positive selected alleles probabilty.</li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of positive fixation.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/fixations.jl#L54-L69">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.fixPosSim" href="#Analytical.fixPosSim"><code>Analytical.fixPosSim</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">fixPosSim(gamma,ppos)</code></pre><p>Expected positive fixations rate reduced due to the impact of background selection and linkage. The probabilty of fixation of positively selected alleles is reduced by a factor Φ across all deleterious linked sites <a href="@ref"><code>Analytical.phiReduction</code></a>.</p><div>\[\mathbb{E}[D_{n+}] =  \Phi \cdot \mathbb{E}[D_{n+}]\]</div><p><strong>Arguments</strong></p><ul><li><code>ppos::Float64</code>: positive selected alleles probabilty.</li></ul><p><strong>Returns</strong></p><ul><li><code>Float64</code>: expected rate of positive fixations under background selection</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/fixations.jl#L91-L107">source</a></section></article><h2 id="Polymorphism-1"><a class="docs-heading-anchor" href="#Polymorphism-1">Polymorphism</a><a class="docs-heading-anchor-permalink" href="#Polymorphism-1" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Analytical.DiscSFSNeutDown" href="#Analytical.DiscSFSNeutDown"><code>Analytical.DiscSFSNeutDown</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">DiscSFSNeutDown()</code></pre><p>Expected rate of neutral allele frequency reduce by background selection. The spectrum depends on the number of individual []</p><div>\[\mathbb{E}[Ps_{(x)}] = \sum{x^{*}=x}{x^{*}=1}f_{B}(x)\]</div><p><strong>Return:</strong></p><ul><li><code>Array{Float64}</code>: expected rate of neutral alleles frequencies.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/polymorphism.jl#L10-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.DiscSFSSelPosDown" href="#Analytical.DiscSFSSelPosDown"><code>Analytical.DiscSFSSelPosDown</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">DiscSFSSelPosDown(gammaValue,ppos)</code></pre><p>Expected rate of positive selected allele frequency reduce by background selection. The spectrum depends on the number of individuals.</p><p><strong>Arguments</strong></p><ul><li><code>gammaValue::Int64</code>: selection strength.</li><li><code>ppos::Float64</code>: positive selected alleles probabilty.</li></ul><p><strong>Return:</strong></p><ul><li><code>Array{Float64}</code>: expected positive selected alleles frequencies.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/polymorphism.jl#L45-L56">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.DiscSFSSelNegDown" href="#Analytical.DiscSFSSelNegDown"><code>Analytical.DiscSFSSelNegDown</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">DiscSFSSelNegDown(param,ppos)</code></pre><p>Expected rate of positive selected allele frequency reduce by background selection. Spectrum drawn on a gamma DFE. It depends on the number of individuals.</p><p><strong>Arguments</strong></p><ul><li><code>ppos::Float64</code>: positive selected alleles probabilty.</li></ul><p><strong>Return:</strong></p><ul><li><code>Array{Float64}</code>: expected negative selected alleles frequencies.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/polymorphism.jl#L159-L169">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.cumulativeSfs" href="#Analytical.cumulativeSfs"><code>Analytical.cumulativeSfs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">cumulativeSfs(sfsTemp)</code></pre><p>Changing SFS considering all values above a frequency <em>x</em>. The original asymptotic-MK approach takes Pn(x) and Ps(x) as the number of polymorphic sites at frequency <em>x</em> rather than above <em>x</em>, but this approach scales poorly as sample size increases. We define the polymorphic spectrum as stated above since these quantities trivially have the same asymptote but are less affected by changing sample size.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/polymorphism.jl#L203-L207">source</a></section></article><h2 id="Summary-statistics-1"><a class="docs-heading-anchor" href="#Summary-statistics-1">Summary statistics</a><a class="docs-heading-anchor-permalink" href="#Summary-statistics-1" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Analytical.poissonFixation" href="#Analytical.poissonFixation"><code>Analytical.poissonFixation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">poissonFixation(observedValues,λds, λdn)</code></pre><p>Divergence sampling from Poisson distribution. The expected neutral and selected fixations are subset through their relative expected rates (<a href="#Analytical.fixNeut"><code>Analytical.fixNeut</code></a>, <a href="#Analytical.fixNegB"><code>Analytical.fixNegB</code></a>, <a href="#Analytical.fixPosSim"><code>Analytical.fixPosSim</code></a>). Empirical values are used are used to simulate the locus <em>L</em> along a branch of time <em>T</em> from which the expected <em>Ds</em> and <em>Dn</em> raw count estimated given the mutation rate (<span>$\mu$</span>). Random number generation is used to subset samples arbitrarily given the success rate <span>$\lambda$</span> in the distribution.</p><div>\[\mathbb{E}[D_N] = X \in Poisson\left(\lambda = D \times \left[\frac{\mathbb{E}[D_+] + \mathbb{E}[D_-]}{\mathbb{E}[D_+] + \mathbb{E}[D_-] + \mathbb{E}[D_0]}\right]\right)\]</div><div>\[\mathbb{E}[D_S] = X \in Poisson\left(\lambda = D \times \left[\frac{\mathbb{E}[D_0]}{\mathbb{E}[D_+] + \mathbb{E}[D_-] + \mathbb{E}[D_0]}\right]\right)\]</div><p><strong>Arguments</strong></p><ul><li><code>observedValues::Array{Int64,1}</code>: Array containing the total observed divergence.</li><li><code>λds</code>: expected neutral fixations rate.</li><li><code>λdn</code>: expected selected fixations rate.</li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Int64,1}</code> containing the expected count of neutral and selected fixations.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/summaryStatistics.jl#L4-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.poissonPolymorphism" href="#Analytical.poissonPolymorphism"><code>Analytical.poissonPolymorphism</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">poissonPolymorphism(observedValues,λps,λpn)</code></pre><p>Polymorphism sampling from Poisson distributions. The total expected neutral and selected polimorphism are subset through the relative expected rates at the frequency spectrum (<a href="#Analytical.fixNeut"><code>Analytical.fixNeut</code></a>, <a href="#Analytical.DiscSFSNeutDown"><code>Analytical.DiscSFSNeutDown</code></a>,). Empirical sfs are used to simulate the locus <em>L</em> along a branch of time <em>T</em> from which the expected <em>Ps</em> and <em>Pn</em> raw count are estimated given the mutation rate (<span>$\mu$</span>). Random number generation is used to subset samples arbitrarily from the whole sfs given each frequency success rate <span>$\lambda$</span> in the distribution.</p><p>The success rate managing the Poisson distribution by the observed count each frequency.  We considered both sampling variance and process variance is affecting the number of variable alleles we sample from SFS. This variance arises from the random mutation-fixation process along the branch. To incorporate this variance we do one sample per frequency-bin and use the total sampled variation and the SFS for the summary statistics.</p><div>\[\mathbb{E}[P_N] = \sum_{x=0}^{x=1} X \in Poisson\left(\lambda = SFS_{(x)} \times \left[\frac{\mathbb{E}[P_{+(x)}] + \mathbb{E}[P_{-(x)}]}{\mathbb{E}[P_{+(x)}] + \mathbb{E}[P_{-(x)}] + \mathbb{E}[P_{0(x)}]}\right]\right)\]</div><div>\[\mathbb{E}[P_S] = \sum_{x=0}^{x=1} X \in Poisson\left(\lambda = SFS_{(x)} \times \left[\frac{\mathbb{E}[P_{0(x)}]}{\mathbb{E}[P_{+(x)}] + \mathbb{E}[P_{-(x)}] + \mathbb{E}[P_{0(x)}]}\right]\right)\]</div><p><strong>Arguments</strong></p><ul><li><code>observedValues::Array{Int64,1}</code>: Array containing the total observed divergence.</li><li><code>λps</code>: expected neutral site frequency spectrum rate.</li><li><code>λpn</code>: expected selected site frequency spectrum rate.</li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Int64,1}</code> containing the expected total count of neutral and selected polymorphism.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/summaryStatistics.jl#L40-L62">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.alphaByFrequencies" href="#Analytical.alphaByFrequencies"><code>Analytical.alphaByFrequencies</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">alphaByFrequencies(gammaL,gammaH,pposL,pposH,observedData,nopos)</code></pre><p>Analytical α(x) estimation. We used the expected rates of divergence and polymorphism to approach the asympotic value accouting for background selection, weakly and strong positive selection. α(x) can be estimated taking into account the role of positive selected alleles or not. We solve α(x) from empirical observed values. The values will be use to sample from a Poisson distribution the total counts of polymorphism and divergence using the rates. The mutation rate, the locus length and the time of the branch should be proportional to the observed values.</p><div>\[\mathbb{E}[\alpha_{x}] =  1 - \left(\frac{\mathbb{E}[D_{s}]}{\mathbb{E}[D_{N}]}\frac{\mathbb{E}[P_{N}]}{\mathbb{E}[P_{S}]}\right)\]</div><p><strong>Arguments</strong></p><ul><li><code>gammaL::Int64</code>: strength of weakly positive selection</li><li><code>gammaH::Int64</code>: strength of strong positive selection</li><li><code>pposL</code>::Float64: probability of weakly selected allele</li><li><code>pposH</code>::Float64: probability of strong selected allele</li><li><code>observedData::Array{Any,1}</code>: Array containing the total observed divergence, polymorphism and site frequency spectrum.</li><li><code>nopos::String(&quot;pos&quot;,&quot;nopos&quot;,&quot;both&quot;)</code>: string to perform α(x) account or not for both positive selective alleles.</li></ul><p><strong>Returns</strong></p><ul><li><code>Tuple{Array{Float64,1},Array{Float64,2}}</code> containing α(x) and the summary statistics array (Ds,Dn,Ps,Pn,α).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/summaryStatistics.jl#L211-L230">source</a></section></article><h2 id="Inference-tools-1"><a class="docs-heading-anchor" href="#Inference-tools-1">Inference tools</a><a class="docs-heading-anchor-permalink" href="#Inference-tools-1" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Analytical.parseSfs" href="#Analytical.parseSfs"><code>Analytical.parseSfs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">parseSfs(;data,output,sfsColumns,divColumns)</code></pre><p>Function to parse polymorphism and divergence by subset of genes. The input data is based on supplementary material described at <a href="https://doi.org/10.1038/s41559-019-0890-6">Uricchio et al. 2019</a>.</p><table><tr><th style="text-align: right">GeneId</th><th style="text-align: right">Pn</th><th style="text-align: right">DAF seppareted by commas</th><th style="text-align: right">Ps</th><th style="text-align: right">DAF separated by commas</th><th style="text-align: right">Dn</th><th style="text-align: right">Ds</th></tr><tr><td style="text-align: right">XXXX</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">0</td></tr><tr><td style="text-align: right">XXXX</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">0</td></tr><tr><td style="text-align: right">XXXX</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">,0,0,0,0,0,0,0,0</td><td style="text-align: right">0</td><td style="text-align: right">0</td></tr></table><p><strong>Arguments</strong></p><ul><li><code>data</code>: String or Array of strings containing files names with full path.</li><li><code>output::String</code>: path to save file. Containing one file per input file.</li><li><code>sfsColumns::Array{Int64,1}</code>: non-synonymous and synonymous daf columns. Please introduce first the non-synonymous number.</li><li><code>divColumns::Array{Int64,1}</code>: non-synonymous and synonymous divergence columns. Please introduce first the non-synonymous number.</li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Array{Int64,N} where N,1}</code>: Array of arrays containing the total polymorphic sites (1), total Site Frequency Spectrum (2) and total divergence (3). Each array contains one row/column per file.</li><li>File writed in <code>output</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/inferTools.jl#L1-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.ABCreg" href="#Analytical.ABCreg"><code>Analytical.ABCreg</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">ABCreg(data, prior, nparams, nsummaries, outputPath, outputPrefix,tolerance, regressionMode,regPath)</code></pre><p>Julia to execute <em>ABCreg</em>. You should take into account that any other ABC could be used once the prior distributions are done. If you consider to use ABCreg please <a href="https://doi.org/10.1186/1471-2156-10-35">cite the publication</a> and compile it in your system.</p><p><strong>Arguments</strong></p><ul><li><code>data::String</code>: Observed data. Produced by parseSfs.</li><li><code>output::String</code>: path to save file. ABCreg will produce one file per lines inside data.</li><li><code>nparams::Int64</code>: number of parameters in prior distributions.</li><li><code>nsummaries::Int64</code>: number of observed summary statistics.</li><li><code>outputPath::String</code>: output path.</li><li><code>outputPrefix::String</code>: output prefix name.</li><li><code>tolerance::Float64</code>: tolerance for rejection acceptance in Euclidian distance.</li><li><code>regressionMode::String</code>: transformation (T or L).</li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Array{Int64,N} where N,1}</code>: Array of arrays containing the total polymorphic sites (1), total Site Frequency Spectrum (2) and total divergence (3). Each array contains one row/column per file.</li><li>File writed in <code>output</code></li><li></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/inferTools.jl#L95-L113">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Analytical.meanQ" href="#Analytical.meanQ"><code>Analytical.meanQ</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia">meanQ(x,columns)</code></pre><p>Function to retrieve mean and quantiles (95%) from posterior distributions.</p><p><strong>Arguments</strong></p><ul><li><code>x::Array{Float64,2}</code>: posterior distribution.</li><li><code>columns::Array{Int64,1}</code>: columns to process.</li></ul><p><strong>Returns</strong></p><ul><li><code>Array{Array{Float64,2},1}</code>: Array of array containing mean and quantiles by posterior distribution. Each array contains <code>$\alpha_{S}$</code>, <code>$\alpha_{W}$</code> and <code>$\alpha$</code> information by column.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jmurga/Analytical.jl/blob/5d6a0f9d33705ef0ff2b3ce5e67ba0e41b5f169e/src/inferTools.jl#L130-L140">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../abc/">« ABC inference</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Tuesday 6 October 2020 11:48">Tuesday 6 October 2020</span>. Using Julia version 1.5.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
